cabal.haskell-ci

```diff
@@ -5,13 +5,5 @@ irc-channels:           irc.freenode.org#haskell-lens
 irc-if-in-origin-repo:  True
 -- Work around https://github.com/haskell/haddock/issues/242
 haddock:                <7.6 || >=7.8
-
--- Because we use a custom Setup script, --test-show-details can only support
--- the options that come with the version of Cabal that is linked against.
--- Unfortunately, --test-show-details=direct wasn't introduced until
--- Cabal-1.24.*, and GHC 7.10 and earlier ship with an older version of Cabal
--- than that. As a result, trying to pass --test-show-details=direct won't work
--- on these GHC versions unless we rebuild Cabal from scratch, which is
--- time-consuming. As a result, we simply disable --test-show-details=direct
--- as a quick-but-dirty workaround.
-test-output-direct:     False
+-- :{ ... :} are problematic for older GHCis
```

Comment:
Why hasn't this been a problem for older GHCs in the current `doctest`-based configuration?

Reply:
Because `doctest` doesn't use `ghci` (but rather reimplements it), and thus can omit printing prompts to begin with.

Reply:
I'll try to find a solution, but I haven't digged in the GHC issue tracker for hints.

Note: with `.ghc.environment.*` approach we could support only GHC-8.x+, as older don't have package-env functionality. So this is already a (small) improvement.

(One way is to not use `:{`... but in `comonad` case that would make examples unreadable)

Reply:
Original issue is https://gitlab.haskell.org/ghc/ghc/-/issues/7509, there doesn't seem to be any easy workarounds to change (i.e. supress) prompt.

Reply:
> Because `doctest` doesn't use `ghci` (but rather reimplements it), and thus can omit printing prompts to begin with.

In what sense does `doctest` re-implement GHCi? My understanding of `doctest` (based on the implementation of [this module](https://github.com/sol/doctest/blob/1bcbe0675694269697f22e75013f770620805bb0/ghci-wrapper/src/Language/Haskell/GhciWrapper.hs) and its use of commands like [`:reload`](https://github.com/sol/doctest/blob/1bcbe0675694269697f22e75013f770620805bb0/src/Runner.hs#L139)) is that is _does_ use GHCi. But perhaps there's a distinction between `doctest`-style GHCi and `docspec`-style GHCi that I'm missing.

> Note: with `.ghc.environment.*` approach we could support only GHC-8.x+, as older don't have package-env functionality. So this is already a (small) improvement.

Now that you mention, I just realized that I have no idea how `cabal-docspec` even works on pre-8.0 versions of GHC that lack the ability to recognize `.ghc.environment.*` files. How _does_ this work currently?

Reply:
> Now that you mention, I just realized that I have no idea how cabal-docspec even works on pre-8.0 versions of GHC that lack the ability to recognize .ghc.environment.* files. How does this work currently?

`cabal-docspec` reads `plan.json` and infers information from there. `plan.json` /= `.ghc.environment.*`.

Reply:
> In what sense does doctest re-implement GHCi? My understanding of doctest (based on the implementation of this module and its use of commands like :reload) is that is does use GHCi. But perhaps there's a distinction between doctest-style GHCi and docspec-style GHCi that I'm missing.

For some reason I thought `doctest` also uses `ghc` lib to implement `ghci`. It doesn't. And somehow specifying `-v0` seems to suppress prompts. But I don't understand how. I have to read GHC manual (or source) to understand whether

```
% echo "1+1" | ghci -v0
Loaded package environment from /home/ogre/Documents/public-haskell/cabal-extras/.ghc.environment.x86_64-linux-8.8.4
2
```

```
% echo "1+1" | ghci    
GHCi, version 8.8.4: https://www.haskell.org/ghc/  :? for help
Loaded package environment from /home/ogre/Documents/public-haskell/cabal-extras/.ghc.environment.x86_64-linux-8.8.4
Loaded GHCi configuration from /home/ogre/.ghci
Prelude> 2
Prelude> Leaving GHCi.
```

is by design.

Let me make a new version of `cabal-docspec` which should work for all GHCs for comonad.

Reply:
It's deliberate; https://gitlab.haskell.org/ghc/ghc/-/blob/cbc7c3dda6bdf4acb760ca9eb545faeb98ab0dbe/ghc/GHCi/UI.hs#L688-691